Cython/Compiler/Nodes.py

```diff
@@ -3885,10 +3901,24 @@ def generate_keyword_unpacking_code(self, min_positional_args, max_positional_ar
             code.putln('default:')
         for i in range(max_positional_args-1, -1, -1):
             code.put('case %2d: ' % (i+1))
-            code.putln("values[%d] = PyTuple_GET_ITEM(%s, %d);" % (
-                i, Naming.args_cname, i))
-            code.putln('CYTHON_FALLTHROUGH;')
-        code.putln('case  0: break;')
+            if i+1 > num_required_posonly_args:
+                code.putln("values[%d] = PyTuple_GET_ITEM(%s, %d);" % (
+                    i, Naming.args_cname, i))
+                code.putln('CYTHON_FALLTHROUGH;')
+            elif i+1 == num_required_posonly_args:
+                for j in range(num_required_posonly_args-1, -1, -1):
+                    code.putln("values[%d] = PyTuple_GET_ITEM(%s, %d);" % (
+                        j, Naming.args_cname, j))
+                code.putln('break;')
+            else:
+                code.putln('CYTHON_FALLTHROUGH;')
+
+        if num_required_posonly_args == 0:
+            code.putln('case  0: break;')
+        else:
+            # catch-all for not enough pos-only args passed
+            code.putln('case  0:')
+            code.put_goto(argtuple_error_label)
```

Comment:
You could generate a one-liner here, in which case you can extract the `code.put('case 0: ')` from the conditional to make it clear that it's always handled.

Reply:
Done